### ISSUE TITLE
NaN and inf checks in PyTorch's actions' implementation

### ISSUE THREAD

**Issue by seovchinnikov** [NVIDIA/NeMo#260]
There is NaN [check](https://github.com/NVIDIA/NeMo/blob/a0c60bcfa2b77c253a7a38d1a33c7ab4efebf94e/nemo/nemo/backends/pytorch/actions.py#L1492) for tensors to optimize.
But some loss functions (e.g. ctc) return inf in some cases and not NaN. But this case is not handled and it leads to [NaN in gradients](https://github.com/pytorch/pytorch/issues/14335) after backward pass and training can not be recovered from there. 
Should we add additional check for inf there and in other places below?

**Comment by seovchinnikov** [comment]
https://github.com/NVIDIA/NeMo/pull/262

**Comment by dmitrytyrin** [comment]
@seovchinnikov, I got pytorch warning "NaN or Inf found in input tensor" at the end of the 0 epoch during asr training. Evaluation Loss is inf. After some steps I got "Warning: Loss is NaN" and training process seemed to crash and stopped to save checkpoints (I got this warning at every step).
My dataset has only characters specified in config file.

Is my issue connected to #262? Is it already fixed? I installed my environment last week.
Also, please check this discussion: https://discuss.pytorch.org/t/warning-nan-or-inf-found-in-input-tensor/48628/4. 

Should I reinstall my environment? Can I use nemo_asr and nemo_toolkit from pypi.org?

Environment: Ubuntu 18.04 with single GPU, quartznet15×5.yaml, quartznet.py, nemo libs from pypi.org.

**Comment by seovchinnikov** [comment]
Your issue is related to this bugfix - if on some iteration your loss is inf (for example because something wrong with CTC for this batch),  training hangs because gradients become NaN.
This bugfix helps to skip batch if it contains inf loss element.

**Comment by okuchaiev** [comment]
thanks @seovchinnikov ! #262 is now merged. Closing this issue

**Comment by dmitrytyrin** [comment]
Dear @okuchaiev and @seovchinnikov,
I've observed interesting case while training ASR model:

Some of my audio (not all dataset) seems to be "broken". I don't know how to find them and what could be wrong with them. My guess that on these audios model gives length predictions (predicted sentence) < than length target (sentence from transcribe). This leads to NaN from CTC loss (please see https://discuss.pytorch.org/t/ctcloss-performance-of-pytorch-1-0-0/27524/10)

I found workaround: use O1 optimisation from APEX and set lr to 0.001. On broken batches I get "Gradient overflow. Skipping step, loss scaler 0 reducing loss scale to 1.0" but training continues and I can perform infer on saved checkpoints. With other settings I get "NaN or inf loss" after first batch with "broken" audios and everything crashes.

1) Is it possible to add "zero_infinity=True" setting to CTC loss (https://github.com/pytorch/pytorch/pull/16199) ?
2) Is it possible add some dummy values to predicted sentence if length(prediction)<length(target) before CTC loss?

**Comment by okuchaiev** [comment]
@dmitrytyrin is this still an issue? If yes, can you please open an issue for it for visibility and with repo steps?

**Comment by dmitrytyrin** [comment]
@okuchaiev, I install latest NeMo (from GitHub) and seems that errors are gone! I still get warning about inf/NaN loss but very rarely and everything works fine with O0 and lr=0.01.
Sorry for inconvenience.
